backend/data_preparation/crawler/fire_crawler.py

```diff
@@ -0,0 +1,318 @@
+        """
+        Takes a year, and a state, fetches all fire events in that year and the state and output it as a list of
+        FireEvent objects.
+        :param year: int, represent the year to crawl.
+                e.g. 2017
+        :param state: str, the state to crawl.
+                e.g. 'California'
+        :return: list of FireEvent objects of all fire events happened in that year this state
+                e.g. [FireEvent(-1, 2015, 'California', 'FireA'), FireEvent(-1, 2015, 'California', 'FireB')]
+        """
+        logger.info(f"Fetching fire events in {state} in year {year}.")
+        current_year = datetime.datetime.now().date().year
+        url_year = "current_year" if current_year == year else year
+        fire_events_page_in_state = FireCrawler._get_url(f"{FireCrawler.BASE_DIR}{url_year}_fire_data/{state}",
+                                                         f"{state} State of year {year}",
+                                                         "_extract_fire_events_in_state()")
+        fire_names_in_this_state = FireCrawler.RE_EXTRACT_FIRE_EVENTS_IN_STATE.findall(fire_events_page_in_state)
+        logger.info(f"Finished fetching fire events in {state} in year {year}.")
+        return list(map(lambda single_fire_name: FireEvent(year, state, single_fire_name), fire_names_in_this_state))
+
+    @staticmethod
+    def _filter_out_invalid_fire_events(fires: List[FireEvent]):
+        """
+        Find those invalid FireEvent objects and remove them from the input list
+        :param fires: list of FireEvent objects.
+                e.g. [FireEvent(-1, 2015, 'California', 'FireA'), FireEvent(-1, 2015, 'California', 'FireB')]
+        """
+        for fire_event in fires[:]:
+            if fire_event.is_valid():
+                fires.remove(fire_event)
+
+    def extract_all_fires(self, start_year: int) -> List[FireEvent]:
+        """
+        Extracts all fires on the website and return.
+        The returned value is a list of tuples:(year:int, state:string, fire's name in its url)
+        :param start_year: int, start year for the crawler. Record before the start year will be deleted.
+                e.g. 2015
+        :return: list of FireEvent objects
+                e.g. [FireEvent(-1, 2015, 'California', 'FireA'), FireEvent(-1, 2015, 'California', 'FireB')]
+        """
+        logger.info("Attempting to get all links...")
+        current_year = datetime.datetime.now().date().year
+        # fetch all year nodes from the website, may encounter errors
+        # add a try-except block, and while loop to keep trying
+        # if attempts > 3 raise an exception
+        try:
+            main_page = self._get_url(FireCrawler.BASE_DIR, "main_page", "extract_all_fires()")
+        except requests.exceptions.RequestException:
+            logger.error("Error: Cannot fetch main_page in function extract_all_fires()")
```

Comment:
no need to re-log here. only log when you catch and handles an error. here I assume you are re-raising the error for outer-scope to handle.